readthedocs/projects/tasks.py

```diff
@@ -87,6 +89,87 @@
 log = logging.getLogger(__name__)
 
 
+class CachedEnvironmentMixin:
+
+    """Mixin that pull/push cached environment to storage."""
+
+    def pull_cached_environment(self):
+        if not self.project.has_feature(feature_id=Feature.CACHED_ENVIRONMENT):
+            return
+
+        storage = get_storage_class(settings.RTD_BUILD_ENVIRONMENT_STORAGE)()
+        filename = self.version.get_storage_environment_cache_path()
+
+        msg = 'Checking for cached environment'
+        log.debug(
+            LOG_TEMPLATE,
+            {
+                'project': self.project.slug,
+                'version': self.version.slug,
+                'msg': msg,
+            }
+        )
+        if storage.exists(filename):
+            msg = 'Pulling down cached environment from storage'
+            log.info(
+                LOG_TEMPLATE,
+                {
+                    'project': self.project.slug,
+                    'version': self.version.slug,
+                    'msg': msg,
+                }
+            )
+            tmp_filename = tempfile.mktemp(suffix='.tar')
+            remote_fd = storage.open(filename, mode='rb')
+            with open(tmp_filename, mode='wb') as local_fd:
+                    local_fd.write(remote_fd.read())
+
+            tar = tarfile.TarFile(tmp_filename)
+            tar.extractall(self.version.project.doc_path)
+
+
+    def push_cached_environment(self):
+        if not self.project.has_feature(feature_id=Feature.CACHED_ENVIRONMENT):
+            return
+
+        project_path = self.project.doc_path
+        paths = [
+            os.path.join(project_path, 'checkouts', self.version.slug),
+            os.path.join(project_path, 'envs', self.version.slug),
+            os.path.join(project_path, 'conda', self.version.slug),
+            os.path.join(project_path, '.cache'),
+        ]
+
+        tmp_filename = tempfile.mktemp(suffix='.tar')
+        # open just with 'w', to not compress and waste CPU cycles
+        with tarfile.open(tmp_filename, 'w') as tar:
+            for path in paths:
+                if os.path.exists(path):
+                    tar.add(
```

Comment:
Noted in chat, we will want to be defensive on tar operations, as we want to preserve and extended attributes and special file types (symlinks especially) on tar operations. A good place to start with tar for full system backups is `tar --xattrs -czpvf ...` -- so maybe the equivalent of this call? If you research full system backups using tar, you'll come across some recommendations i'm sure.

Reply:
I didn't find anything particularly useful. The recommended attributes to create a full system backup are the defaults in Python. So, I think we are good here.